clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1176,18 +1176,25 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
                 metadata.awaitUpdate(version, remainingWaitMs);
             } catch (TimeoutException ex) {
                 // Rethrow with original maxWaitMs to prevent logging exception with remainingWaitMs
-                throw new TimeoutException(
-                        String.format("Topic %s not present in metadata after %d ms.",
-                                topic, maxWaitMs));
+                final String errorMessage = String.format("Topic %s not present in metadata after %d ms.",
+                        topic, maxWaitMs);
+                if (metadata.getError(topic) != null) {
+                    throw new TimeoutException(errorMessage, metadata.getError(topic).exception());
+                }
+                throw new TimeoutException(errorMessage);
             }
             cluster = metadata.fetch();
             elapsed = time.milliseconds() - nowMs;
             if (elapsed >= maxWaitMs) {
-                throw new TimeoutException(partitionsCount == null ?
+                final String errorMessage = partitionsCount == null ?
                         String.format("Topic %s not present in metadata after %d ms.",
                                 topic, maxWaitMs) :
                         String.format("Partition %d of topic %s with partition count %d is not present in metadata after %d ms.",
-                                partition, topic, partitionsCount, maxWaitMs));
+                                partition, topic, partitionsCount, maxWaitMs);
+                if (metadata.getError(topic) != null) {
```

Comment:
Here the metadata update did happen (so we may have received errors in the response), but still, it took longer than our `maxWaitMs` so we want to throw a timeout instead. So in this case, do we really want to show an InvalidTopic/UnauthorizedTopic/... error as the root cause for the timeout?  (Sounds a bit confusing to me as "cause", but I totally may be missing how metadata errors like these made us take longer and timeout).  If we had received the response in time, those errors would be propagated in the following `maybeThrowExceptionForTopic`, but here we choose timeout over the errors right?

Reply:
Thanks for the clarification, @lianetm. So, the following cases may happen:
1) a timeout and a non-retriable error (such as InvalidTopic/UnauthorizedTopic) happen
2) a timeout and a retriable error (such as UnknownTopicOrPartition) happen
3) only a non-retriable error happens
- Your worries are about case (1), where non-retriable errors are listed as the cause of the timeout exception, which is itself retriable.
- The purpose of this PR, is to actually know the cause of the timeout exceptions in case (2) to prevent complications such as the one explained in [KAFKA-16508](https://issues.apache.org/jira/browse/KAFKA-16508)
- Case (3) is managed by `maybeThrowExceptionForTopic` and I think we both have no worries about that.

Please correct me if I am wrong or incomplete in listing the cases. If I'm right, changing the code in a way that just retriable errors get listed as the cause of timeout exception ([here](https://github.com/aliehsaeedii/kafka/blob/a27a458d03a173691053c21ebcfb21f11d3cad12/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L1195)) would be easy.

Reply:
Would we see case 1? A timeout and a non-retriable error? If we receive a non-retriable error we don't care about timeouts anymore right?

Reply:
@jolshan agree we don't care about timeout if we got a fatal error, but with the changes I was concerned we could get case 1: edge case where we kept retrying, and on the last try before timing out, we got a non-retriable. In that case, our logic here would throw a timeout, but including the the non-retriable as cause right?

Other than this edge case, agree we shouldn't get 1, fatal errors would get propagated on the `maybeThrowExceptionForTopic` a few lines down, all good. 

So seems we are all aligned on reducing the scope to include the "last known error" as cause of the timeout here, but only if it's a retriable one. Sounds good to me to solve what we wanted without messing up the flow for non-retriables, that should remain unchanged I expect.

Reply:
I see. I guess I wasn't sure if we would get to this flow if we timed out with a non-retriable error. (Do we still have this timer if we hit a non-retriable error)

Reply:
yes, it's tricky but we do have it mainly because the flow checks elapsed time before throwing fatals (not the opposite), so we could have an expired "timer" with a fatal in hand (that we haven't thrown just yet).

Reply:
Interesting. I wonder if that is that is the only way. But we don't need to change now I suppose.

Reply:
Interesting -- I would also think that checking for fatal error should be done first (if possible) -- but we should not block this PR on this question either -- maybe, if we file a ticket for updating the log, we could point out the changes in this PR and simplify the code we add here, if this corner case goes away.